contracts/pcv/sentinel/PCVSentinel.sol

```
@@ -0,0 +1,173 @@
+                numGuardsCanProtec++;
+            }
+        }
+
+        address[] memory guardsWhoCanProtec = new address[](numGuardsCanProtec);
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec--;
+                guardsWhoCanProtec[numGuardsCanProtec] = guards.at(i);
+            }
+        }
+
+        return guardsWhoCanProtec;
+    }
+
+    // ---------- Governor-or-Admin-Or-Guardian-Only State-Changing API ----------
+
+    /**
+     * @notice adds a guard contract to the PCV Sentinel
+     * @param guard the guard-contract to add
+     */
+    function knight(address guard) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.add(guard);
+
+        // Inform the kingdom of this glorious news
+        emit GuardAdded(guard);
+    }
+
+    /**
+     * @notice removes a guard
+     * @param traitor the guard-contract to remove
+     */
+    function slay(address traitor) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.remove(traitor);
+
+        // Inform the kingdom of this sudden and inevitable betrayal
+        emit GuardRemoved(traitor);
+    }
+
+    // ---------- Public State-Changing API ----------
+
+    /**
+     * @notice checks the guard and runs its protec actions if needed
+     * @param guard the guard to activate
+     */
+    function protec(address guard) 
+        external
+        override
+    {
+        require(guards.contains(guard), "Guard does not exist.");
+
+        bool needsProtec = IGuard(guard).check();
+
+        if (needsProtec) {
+            (address[] memory targets, bytes[] memory datas) = IGuard(guard).getProtecActions();
+
+            for(uint256 i=0; i<targets.length; i++) {
+                require(targets[i] != address(this), "Nyeh!");
+                targets[i].call(datas[i]); // Unopinionated about reverts
```

What do you mean by "Unopinionated about reverts" ?

---

the pcv sentinel does not care if the actions it takes are reverting. adds complexity; the guard can handle that.

---

I actually feel pretty strongly we should fail loud